pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,353 @@
+package blockfetcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+
+	"github.com/nspcc-dev/neo-go/pkg/config"
+	"github.com/nspcc-dev/neo-go/pkg/core/block"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	gio "github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/network/bqueue"
+	"github.com/nspcc-dev/neo-go/pkg/network/payload"
+	"github.com/nspcc-dev/neo-go/pkg/services/oracle/neofs"
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+	"github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"go.uber.org/zap"
+)
+
+const (
+	ModeIndexSearch = "indexSearch"
+	ModeOidSearch   = "oidSearch"
+)
+
+// Ledger is an interface to Blockchain sufficient for Service.
+type Ledger interface {
+	AddBlock(block *block.Block) error
+	GetConfig() config.Blockchain
+	BlockHeight() uint32
+	AddHeaders(...*block.Header) error
+	HeaderHeight() uint32
+}
+
+// Service is a service that fetches blocks from NeoFS.
+type Service struct {
+	chain      Ledger
+	client     *client.Client
+	log        *zap.Logger
+	quit       chan bool
+	cfg        config.NeoFSBlockFetcher
+	queue      *bqueue.Queue
+	privateKey *keys.PrivateKey
+}
+
+// New creates a new BlockFetcherService.
+func New(chain Ledger, cfg config.NeoFSBlockFetcher, queue *bqueue.Queue, logger *zap.Logger) *Service {
+	privateKey, err := keys.NewPrivateKey()
+	if err != nil {
+		return nil
+	}
+	return &Service{
+		chain:      chain,
+		log:        logger,
+		quit:       make(chan bool),
+		cfg:        cfg,
+		queue:      queue,
+		privateKey: privateKey,
+	}
+}
+
+// Name implements the core.Service interface.
+func (bfs *Service) Name() string {
+	return "BlockFetcherService"
+}
+
+// Start implements the core.Service interface.
+func (bfs *Service) Start() {
+	bfs.log.Info("Starting Block Fetcher Service")
+	c, err := neofs.GetSDKClient(context.Background(), bfs.cfg.Addresses[0])
+	defer func(c *client.Client) {
+		err := c.Close()
+		if err != nil {
+			bfs.log.Error("Failed to close SDK client", zap.Error(err))
+		}
+	}(c)
+
+	if err != nil {
+		bfs.log.Error("Failed to create SDK client", zap.Error(err))
+		close(bfs.quit)
+		return
+	}
+	bfs.client = c
+	switch bfs.cfg.Mode {
+	case ModeIndexSearch:
+		err = bfs.fetchData()
+	case ModeOidSearch:
+		err = bfs.fetchDataWithOid()
+	default:
+		bfs.log.Error("Invalid mode specified", zap.String("mode", bfs.cfg.Mode))
+		close(bfs.quit)
+		return
+	}
+
+	if err != nil {
+		bfs.log.Error("Fetch operation failed", zap.Error(err))
+		close(bfs.quit)
+		return
+	}
+}
+
+func (bfs *Service) GetHeaders() error {
+	startIndex := bfs.chain.HeaderHeight() / 2000
```

Comment:
For now I'm leaving `GetHeaders` without review because we firstly need to pick the block fetching strategy and then reuse the same strategy for headers.